### PR TITLE
[fullstack] Added support for GraphQL

### PR DESCRIPTION
**Pull request by advance512**
Added support for GraphQL:
   Added react-apollo support
   Added client-side Apollo-client
   Added server-side Apollo-client for SSR
   Added GraphQL endpoint and interactive endpoint in server
Updated .babelrc to use latest ES2015/2016/2017 features
Added minilog package and implemented basic logger service and error logger
Added casual for random mocks
Added isomorophic-fetch for better Service Side Rendering
Added GraphQLImage atom and SVG image
Deleted SamplePage component
Added SamplePageUsingAPI, PostListUsingAPI, PostFormUsingAPI Containers
Added SamplePageUsingGraphQL, PostListUsingGraphQL, PostFormUsingGraphQL Containers
Updated PrimaryNavigation with new page
Using devtool: eval-source-map in Webpack for proper debugging

Updated src/ and src-clean/ too

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "presets": ["es2015", "react", "stage-1"],
+  "presets": ["latest", "react", "stage-1"],
```
Comment (diegohaz) [inline .babelrc]: Could you please make this change in a separate PR against the master branch? I already wanted to replace `babel-preset-es2015` for that. :)

File: .gitignore [hunk 2]
```diff
@@ -4,4 +4,7 @@ yarn-error.log
 npm-debug.log
 node_modules
 coverage
-webpack/webpack-assets.json
+webpack/
+.idea/
+generators/
+templates/
```
Comment (diegohaz) [inline .gitignore]: I see that's something we need to do to better handle branch switching. It would be also very desirable in separate PRs.

File: src-clean/config.js [hunk 3]
```diff
@@ -10,6 +10,8 @@ const config = {
     env: process.env.NODE_ENV || 'development',
     baseUrl: `http://${ip}:${port}${basename}`,
     apiUrl: `http://${ip}:${port}/api`,
+    graphqlUrl: `http://${ip}:${port}/api/graphql`,
+    serverGraphqlUrl: 'http://localhost:3000/api/graphql',
```
Comment (diegohaz) [inline src-clean/config.js]: I think we should make `baseUrl` a variable (like `ip` and `port`) and use it for `apiUrl`, `graphqUrl` and `serverGraphqlUrl`. The `basename` is necessary for cases when you need to serve the project in a path other than root (e.g. to serve on `https://diegohaz.github.io/foo` we need to `PUBLIC_PATH=foo npm run build`)

File: src-clean/server.js [hunk 4]
```diff
@@ -1,27 +1,39 @@
 /* eslint-disable no-console */
+
+// Necessary for SSR
+import 'isomorphic-fetch'
```
Comment (diegohaz) [inline src-clean/server.js]: I'm just curious. Could you please explain the need for it?

File: webpack/webpack.config.js [hunk 5]
```diff
@@ -9,7 +9,7 @@ const DEBUG = process.env.NODE_ENV !== 'production'
 const PUBLIC_PATH = `/${process.env.PUBLIC_PATH || ''}/`.replace('//', '/')
 
 const config = {
-  devtool: DEBUG ? 'eval' : false,
+  devtool: DEBUG ? 'eval-source-map' : false,
```
Comment (diegohaz) [inline webpack/webpack.config.js]: Is this better than `eval`?

Comment (Kyezil) [inline webpack/webpack.config.js]: Source-maps helps debugging front-end, makes builds slower though.
https://webpack.github.io/docs/build-performance.html#sourcemaps

File: src-clean/api/index.js [hunk 6]
```diff
@@ -1,5 +1,22 @@
 import { Router } from 'express'
+import { graphiqlExpress, graphqlExpress } from 'graphql-server-express'
+import { schema, mocks } from './graphql'
 
 const router = new Router()
 
+// Interactive GraphQL browser
+export const graphQLRouter = new Router()
+router.use('/graphiql', graphiqlExpress({
+  endpointURL: '/graphql'
```
Comment (Kyezil) [inline src-clean/api/index.js]: Shouldn't this be 'graphql' or '/api/graphql' ?

**Comment by advance512** [comment]
Hi!
Sorry, I've been quite busy with a project I'm working on, didn't get to ping-back here.
Let me look at the comments.

File: .babelrc (hunk 1 above)
Comment (advance512) [inline .babelrc]: Makes sense

File: .gitignore (hunk 2 above)
Comment (advance512) [inline .gitignore]: Makes sense, too :)

File: src-clean/server.js (hunk 4 above)
Comment (advance512) [inline src-clean/server.js]: I'm sure you've [looked it up](https://github.com/matthew-andrews/isomorphic-fetch) by now, but: isometric fetch adds support for fetch both on the client and the server, in the most platform-appropriate way. Many React/Redux libraries require a working fetch(), and doing SSR means you need to use node-fetch - or isometric-fetch which solves the issue for you.

File: webpack/webpack.config.js (hunk 5 above)
Comment (advance512) [inline webpack/webpack.config.js]: We are talking about debug mode. Being able to debug the application is very important for me, but feel free to change this back.

File: src-clean/api/index.js (hunk 6 above)
Comment (advance512) [inline src-clean/api/index.js]: Yes, it should be /api/graphql. This makes me feel like I missed something..

**Comment by Geczy** [comment]
Hi @advance512, any word?

**Comment by advance512** [comment]
Hi there,

I will not be able to take on as a maintainer, not in the near future anyways - my new venture is taking all of my time. My current code has gone very far from its ARc origins, but some common core remains. I use Apollo extensively for just about every page and it's awesome.

I suggest someone else takes over this PR and possibly the other 2:

https://github.com/diegohaz/arc/pull/118
https://github.com/diegohaz/arc/pull/117

If you're interested at all in using GraphQL. :)

**Comment by Geczy** [comment]
@diegohaz do you have interest in doing a graphql branch / example added to the redux branch?

**Comment by diegohaz** [comment]
> do you have interest in doing a graphql branch

Yeah! 👍 

But I will need someone willing to maintain it because I don't use nor have enough knowledge on graphql

**Comment by Geczy** [comment]
I will be using it very often so I'm willing to maintain if you're willing to create :)

**Comment by diegohaz** [comment]
Just ping me on our [chat](https://gitter.im/diegohaz/arc) when you are ready and I'll add you as a colaborator.
